monai/metrics/metric.py

```diff
@@ -0,0 +1,199 @@
+
+        return ret
+
+    def _compute_list(self, y_pred: TensorOrList, y: Optional[TensorOrList] = None):
+        """
+        Excute the computation for every item of a list.
+        Subclass may enhance the operation with multi-threads to accelerate.
+
+        """
+        if y is not None:
+            return [
+                self._compute_tensor(p_.detach().unsqueeze(0), y_.detach().unsqueeze(0)) for p_, y_ in zip(y_pred, y)
+            ]
+        else:
+            return [self._compute_tensor(p_.detach().unsqueeze(0), None) for p_ in y_pred]
+
+    @abstractmethod
+    def _compute_tensor(self, y_pred: torch.Tensor, y: Optional[torch.Tensor] = None):
+        """
+        computation logic for tensors, input data should be `batch-first` Tensor.
+
+        """
+        raise NotImplementedError(f"Subclass {self.__class__.__name__} must implement this method.")
+
+
+class Cumulative:
+    """
+    Utility class for the typical cumulative computation process based on PyTorch Tensor data.
+    Execute the steps referring to below examples::
+
+        cum = Cumulative(buffer_count=2)
+        cum.add(x, y)
+        cum.add(a, b)
+        cum.add(c, d)
+        cum.agrregate()
+        result = cum.get_synced_tensors()
+        cum.reset()
+
+    Args:
+        buffer_num: the number of buffers to create to cumulate PyTorch Tensors,
+            usually every Tensor should map to a buffer.
+
+    """
+
+    def __init__(self, buffer_num: int = 1):
+        self.buffer_num = buffer_num
+        self.reset()
+
+    def reset(self):
+        """
+        Reset buffers to cumulate tensors and the synced results.
+
+        """
+        self._buffers: List[List[torch.Tensor]] = [[] for _ in range(self.buffer_num)]
+        self._synced_tensors: List[Optional[torch.Tensor]] = [None for _ in range(self.buffer_num)]
+
+    def add(self, *data: torch.Tensor):
+        """
+        Add samples to the cumulative buffers.
+
+        Args:
+            data: list of input tensor, make sure the input data order is always the same.
+                every item of data will be added to the corresponding buffer.
+
+        """
+        for i, d in enumerate(data):
+            if not isinstance(d, torch.Tensor):
+                raise ValueError(f"the data to cumulate in a buffer must be PyTorch Tensor, but got: {type(d)}.")
+            self._buffers[i].append(d)
+
+    def aggregate(self):
+        """
+        Aggregate final results based on buffers.
+        This base class only syncs the data from distributed ranks, subclasses should implement more logic.
+
+        """
+        self._sync()
+
+    def _sync(self):
+        """
+        Sync up distributed data when aggregating.
+
+        """
+        self._synced_tensors = [evenly_divisible_all_gather(torch.cat(b, dim=0), concat=True) for b in self._buffers]
+
+    def get_synced_tensors(self):
```

Comment:
this method name could be `read_buffer` and it should call `self._sync`.

Reply:
OK, then it also needs a flag `self._synced: bool = False` to avoid duplicated syncing, considering the case that we may need to read metric details several times to generate reports.

Thanks.

Reply:
Updated in the latest commit.
Thanks.